File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
class Solution {
    private Map<Node, Node> visitedHash = new HashMap<>();

    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }

        if (visitedHash.containsKey(head)) {
            return visitedHash.get(head);
        }

        Node node = new Node(head.val);
        visitedHash.put(head, node);

        node.next = copyRandomList(head.next);
        node.random = copyRandomList(head.random);

        return node;
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.js
function copyRandomList(head) {
    const visitedHash = new Map();

    function clone(node) {
        if (node === null) {
            return null;
        }

        // 이미 복사한 노드라면 기존 복사본 반환
        if (visitedHash.has(node)) {
            return visitedHash.get(node);
        }

        // 새로운 노드 생성
        const newNode = new Node(node.val);

        // 반드시 재귀 호출 전에 저장
        // random/next가 서로를 가리킬 수 있기 때문
        visitedHash.set(node, newNode);

        newNode.next = clone(node.next);
        newNode.random = clone(node.random);

        return newNode;
    }

    return clone(head);
}